virtool/subtractions/api.py

```diff
@@ -173,6 +185,69 @@ async def create(req):
     return json_response(virtool.utils.base_processor(document), headers=headers, status=201)
 
 
+@routes.post("/api/subtractions/{subtraction_id}/files", permission="modify_subtraction")
+async def upload(req):
+    """
+    Upload a new subtraction file to the `subtraction_files` SQL table and the `subtractions` folder in the Virtool
+    data path.
+
+    """
+    db = req.app["db"]
+    pg = req.app["pg"]
+    subtraction_id = req.match_info["subtraction_id"]
+
+    errors = virtool.uploads.utils.naive_validator(req)
+
+    if errors:
+        return invalid_query(errors)
+
+    document = await db.subtraction.find_one(subtraction_id)
+
+    if document is None:
+        return not_found()
+
+    file_name = req.query.get("name")
+
+    if file_name not in FILES:
+        return bad_request("Unsupported subtraction file name")
+
+    if await db.subtraction.count_documents({"_id": subtraction_id, "files.name": file_name}):
+        return bad_request("File name already exists")
+
+    file_type = virtool.subtractions.utils.check_subtraction_file_type(file_name)
+    subtraction_file = await virtool.subtractions.files.create_subtraction_file(pg, subtraction_id, file_type, file_name)
+    file_id = subtraction_file["id"]
+    path = Path(req.app["settings"]["data_path"]) / "subtractions" / subtraction_id / file_name
+
+    try:
+        size = await virtool.uploads.utils.naive_writer(req, path)
+    except asyncio.CancelledError:
+        logger.debug(f"Subtraction file upload aborted: {file_id}")
+        await virtool.subtractions.files.delete_subtraction_file(pg, file_id)
+
+        return aiohttp.web.Response(status=499)
+
+    subtraction_file = await virtool.uploads.db.finalize(pg, size, file_id, SubtractionFile)
+
+    file = {
+        "name": file_name,
+        "size": size,
+        "type": file_type
+    }
+
+    await db.subtraction.find_one_and_update({"_id": subtraction_id}, {
+        "$push": {
+            "files": file
```

Comment:
Store file IDs in this array. They will be used to join the SQL records to the subtraction when it is requested.

Reply:
Only file IDs?